cdap-ui/server/aggregator.js

```diff
@@ -43,59 +49,104 @@ Aggregator.prototype.log = function () {
 };
 
 /**
- * schedule polling
+ * Checks if the 'id' received from the client is already registered -- This
+ * check was added because for whatever reason 'Safari' was sending multiple
+ * requests to backend with same ids. As this is happens only once during
+ * the start of poll, it's safe to make this check. The assumption here is
+ * that the frontend is sending in unique ids within the websocket session.
+ *
+ * Upon check if it's not duplicate, we invoke doPoll that would make the
+ * first call and set the interval for the timeout.
+ */
+Aggregator.prototype.startPolling = function (resource) {
+  // WARN: This assumes that the browser side ids are unique for a websocket session.
+  // This check is needed for Safari.
+  if(this.polledResources[resource.id]) {
+    // console.log("Resource id " + resource.id + " already registered.");
+    return;
+  }
+  resource.interval = resource.interval || POLL_INTERVAL;
+  this.polledResources[resource.id] = resource;
+  _.bind(doPoll, this, resource)();
+}
+
+/**
+ * This method is called regularly by 'doPoll' to register the next interval
+ * for timeout. Every resource handle has a flag is used to indicate if the
+ * the resource has been requested to be stopped, if it's already stopped, then
+ * there is nothing for us to do. If it's not then we go ahead and register
+ * the interval timeout.
  */
-Aggregator.prototype.planPolling = function () {
-  this.timeout = setTimeout(_.bind(doPoll, this), POLL_INTERVAL);
+Aggregator.prototype.scheduleAnotherIteration = function (resource) {
+  if (resource.stop) {
+    // Don't reschedule another iteration if the resource has been stopped
+    return;
+  }
+  // console.log(Object.keys(this.polledResources).length + ':' + Math.floor(Date.now()/1000) +
+  //    ': scheduling for: ' + resource.id + ', interval: ' + resource.interval + ' - ' + resource.url);
+  resource.timerId = setTimeout(_.bind(doPoll, this, resource), resource.interval);
```

Comment:
It's not to avoid sending a response for a resource which has stopped polling.
Rather, its to avoid scheduling another request.